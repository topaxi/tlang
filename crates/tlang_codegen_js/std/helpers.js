/**
 * JavaScript Iterator Helper Functions
 * Provides runtime support for converting JavaScript iterables to tlang iterator protocol
 */

class TlangIteratorWrapper {
  #jsIterator;

  /**
   * Creates a tlang iterator from a JavaScript iterator
   * @param {Iterator<any>} jsIterator - The JavaScript iterator to convert
   * @returns {Object} An iterator object with a next() method that returns Option::Some or Option::None
   */
  static from(jsIterator) {
    return new TlangIteratorWrapper(jsIterator);
  }

  constructor(jsIterator) {
    this.#jsIterator = jsIterator;
  }

  next() {
    const { done, value } = this.#jsIterator.next();

    return done ? Option.None : Option.Some(value);
  }
}

/**
 * Iterator module for tlang
 * Provides a unified interface for creating iterators from different types
 */
const iterator = {
  /**
   * Creates an iterator for the given value
   * Handles anything JavaScript for...of can handle (arrays, strings, Sets, Maps, generators, etc.)
   * @param {any} value - The value to create an iterator for
   * @returns {Object} An iterator object compatible with tlang iterator protocol
   */
  iter: function (value) {
    // Check if the value implements the JavaScript iterator protocol (Symbol.iterator)
    // This handles arrays, strings, Sets, Maps, generators, typed arrays, NodeLists, etc.
    if (value != null && typeof value[Symbol.iterator] === 'function') {
      return TlangIteratorWrapper.from(value[Symbol.iterator]());
    }

    // Check if the object is already an iterator (has a next method)
    if (value != null && typeof value.next === 'function') {
      // Wrap existing iterator to ensure tlang compatibility
      return TlangIteratorWrapper.from(value);
    }

    // Check if it has a tlang iter method and call it
    if (value != null && typeof value.iter === 'function') {
      return value.iter();
    }

    // If none of the above, throw an error as the value is not iterable
    throw new Error(`Value is not iterable: ${typeof value} ${value}`);
  },
};

// Expose iterator for testing/external use
if (typeof globalThis !== 'undefined') {
  globalThis.iterator = iterator;
} else if (typeof window !== 'undefined') {
  window.iterator = iterator;
}
