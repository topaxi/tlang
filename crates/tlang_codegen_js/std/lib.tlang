enum Option {
  Some(value),
  None,
}

fn Option.is_some(Option::Some(_)) { true }
fn Option.is_some(_) { false }

fn Option.is_none(Option::None) { true }
fn Option.is_none(_) { false }

fn Option.unwrap(Option::Some(x)) { x }
fn Option.unwrap(_) { panic("Called unwrap on None") }

enum Result {
  Ok(value),
  Err(error),
}

fn Result.is_ok(Result::Ok(_)) { true }
fn Result.is_ok(_) { false }

fn Result.is_err(Result::Err(_)) { true }
fn Result.is_err(_) { false }

fn Result.unwrap(Result::Ok(x)) { x }
fn Result.unwrap(_) { panic("Called unwrap on Err") }

// len(a[]) -> int
fn len(list) { list.length }

// random_int(int) -> int
fn random_int(max) { math::random() * max |> math::floor() }

// compose(fn(a) -> b, fn(b) -> c) -> fn(a) -> c
fn compose(f, g) { fn(x) { f(g(x)) } }

// map(a[], fn(a) -> b) -> b[]
fn map([], _) { [] }
fn map([x, ...xs], f) { [f(x), ...map(xs, f)] }

// filter(a[], fn(a) -> bool) -> a[]
fn filter([], _) { [] }
fn filter([x, ...xs], f) if f(x) { [x, ...filter(xs, f)] }
fn filter([_, ...xs], f) { rec filter(xs, f) }

// partition(a[], fn(a) -> bool) -> (a[], a[])
fn partition([], _) { [[], []] }
fn partition(list, predicate) { partition(list, predicate, [], []) }
// partition(a[], fn(a) -> bool, a[], a[]) -> (a[], a[])
fn partition([], _, satisfies, does_not_satisfy) { [satisfies, does_not_satisfy] }
fn partition([x, ...xs], predicate, satisfies, does_not_satisfy) if predicate(x) {
  rec partition(xs, predicate, [...satisfies, x], does_not_satisfy)
}
fn partition([x, ...xs], predicate, satisfies, does_not_satisfy) {
  rec partition(xs, predicate, satisfies, [...does_not_satisfy, x])
}

// filter_map(a[], fn(a) -> Option(b)) -> b[]
fn filter_map([], _) { [] }
fn filter_map([x, ...xs], f) if let Option::Some(y) = f(x) { [y, ...filter_map(xs, f)] }
fn filter_map([_, ...xs], f) { rec filter_map(xs, f) }

// foldl(a[], b, fn(b, a) -> b) -> b
fn foldl([], acc, _) { acc }
fn foldl([x, ...xs], acc, f) { rec foldl(xs, f(acc, x), f) }

// foldr(a[], b, fn(a, b) -> b) -> b
fn foldr([], acc, _) { acc }
fn foldr([x, ...xs], acc, f) { f(x, foldr(xs, acc, f)) }

// sum(int[] xs) -> int
fn sum([]) { 0 }
fn sum([x, ...xs]) { x + sum(xs) }

// zip(a[], b[]) -> [a, b][]
fn zip([], []) { [] }
fn zip([x, ...xs], [y, ...ys]) { [[x, y], ...zip(xs, ys)] }
