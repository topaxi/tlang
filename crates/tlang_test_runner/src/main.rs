use std::fs;
use std::path::Path;
use std::process::{Command, Stdio};

use glob::glob;

enum Backend {
    Interpreter,
    JavaScript,
}

impl Backend {
    fn as_str(&self) -> &'static str {
        match self {
            Backend::Interpreter => "interpreter",
            Backend::JavaScript => "javascript",
        }
    }

    fn values() -> impl Iterator<Item = Backend> {
        vec![Backend::Interpreter, Backend::JavaScript].into_iter()
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let pattern = "tests/**/*.tlang";
    let mut errors: Vec<_> = vec![];

    for (i, backend) in Backend::values().enumerate() {
        if i > 0 {
            println!();
        }

        println!("Running tests with backend: {}", backend.as_str());

        for entry in glob(pattern).expect("Failed to read glob pattern") {
            let file_path = entry.expect("Failed to read test file path");

            match run_test(&file_path, backend.as_str()) {
                Ok(()) => {}
                Err(e) => {
                    errors.push(e);
                }
            }
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors.join("\n").into())
    }
}

fn run_test(file_path: &Path, backend: &str) -> Result<(), String> {
    let expected_output_path = {
        let base_path = file_path.with_extension(""); // Strip `.tlang` extension
        let backend_path = base_path.with_extension(format!("expected.{}.txt", backend));
        if backend_path.exists() {
            backend_path
        } else {
            base_path.with_extension("expected.txt") // Fallback to default
        }
    };

    let expected_output = fs::read_to_string(&expected_output_path)
        .map_err(|e| format!("Failed to read {}: {}", expected_output_path.display(), e))?;

    let start = std::time::Instant::now();
    let exec_start;
    let output = match backend {
        "interpreter" => {
            exec_start = std::time::Instant::now();

            Command::new("./target/release/tlangdi")
                .arg(file_path)
                .output()
                .map_err(|e| format!("Failed to execute interpreter: {}", e))?
        }
        "javascript" => {
            #[allow(clippy::zombie_processes)]
            let tlang_js_compiler_output = Command::new("./target/release/tlang_cli_js")
                .arg(file_path)
                .stdout(Stdio::piped())
                .spawn()
                .expect("Failed to compile to JavaScript");

            exec_start = std::time::Instant::now();

            let javascript_output = Command::new("node")
                .stdin(tlang_js_compiler_output.stdout.unwrap())
                .stdout(Stdio::piped())
                .spawn()
                .expect("Failed to execute JavaScript");

            javascript_output.wait_with_output().unwrap()
        }
        _ => return Err(format!("Unsupported backend: {}", backend)),
    };
    let elapsed = start.elapsed();
    let exec_elapsed = exec_start.elapsed();

    let actual_output = if output.status.success() {
        String::from_utf8_lossy(&output.stdout)
    } else {
        String::from_utf8_lossy(&output.stderr)
    };

    if actual_output.trim() == expected_output.trim() {
        println!(
            "Test passed for {} in {:?} (total {:?})",
            file_path.display(),
            exec_elapsed,
            elapsed
        );
        Ok(())
    } else {
        println!(
            "Test failed for {}.\nExpected:\n{}\nActual:\n{}",
            file_path.display(),
            expected_output,
            actual_output
        );
        Err("Test failed".to_string())
    }
}
