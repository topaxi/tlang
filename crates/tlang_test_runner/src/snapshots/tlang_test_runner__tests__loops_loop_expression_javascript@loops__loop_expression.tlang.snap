---
source: crates/tlang_test_runner/src/main.rs
expression: apply_redactions(&output)
input_file: tests/loops/loop_expression.tlang
---

